Migrate classic meetings to dynamic meetings

oliverguenther opened the pull request:
https://community.openproject.org/work_packages/62621

oliverguenther committed:
Allow meeting outcomes to have an author

oliverguenther committed:
Add migration for moving classic meetings

oliverguenther committed:
Remove references to StructuredMeeting, Agenda and Minutes

oliverguenther committed:
Remove content API

oliverguenther committed:
Ensure all agenda items get destroyed

oliverguenther committed:
Close classic meetings in the past

mrmir commented on modules/meeting/app/services/meetings/copy_service.rb:
```diff
@@ -114,31 +114,14 @@ def update_references(attachment_source:, attachment_target:, model_source:, mod
-      end
-    end
-
-    def copy_structured_meeting_participants(copy)
```
👍🏽

mrmir commented on modules/meeting/config/locales/en.yml:
```diff
@@ -82,8 +82,7 @@ en:
       recurring_meeting: "Recurring meeting"
-      structured_meeting: "One-time meeting"
-      meeting: "Meeting"
+      meeting: "One-time Meeting"
```
```suggestion
      meeting: "One-time meeting"
```

mrmir commented on modules/meeting/db/migrate/20250404061503_migrate_classic_meetings.rb:
```diff
@@ -0,0 +1,81 @@
+    # Create default sections for meetings that don't have one
+    execute <<~SQL
+      INSERT INTO meeting_sections (meeting_id, title, position, created_at, updated_at)
+      SELECT m.id, '', 1, NOW(), NOW()
```
I wonder if "Meeting agenda" or something similar can be the default title? They look a bit weird with an empty title

oliverguenther replied:
This is the section which should be empty to signal the default section group. Let's maybe use the existing key "Agenda" because that's most universal, since that title cannot be translated

mrmir replied:
Aah yes my bad, I meant on the agenda item being created

And yeah, "Agenda" is fine for me too, but I hadn't thought of translations. If that is going to be a potential problem, leaving it blank is okay as well. It's just that there's no other way to add a blank titled agenda item, so the behaviour is a bit strange

mrmir commented on modules/meeting/lib/api/v3/attachments/attachments_by_meeting_content_api.rb:
```diff
@@ -32,23 +32,18 @@ module V3
-              api_v3_paths.attachments_by_meeting_content container.id
-            end
+          get do
+            status :gone
```
(For my info) What does this do? Why is it needed?

oliverguenther replied:
To signal to API users, that this endpoint is no longer available without returning a 404 (different semantics)

mrmir reviewed (changes requested):
Hi, there's still a few things that need to be looked at, apart from the minor inline comments:

1. The "+ Meeting" menu still has an option for classic meeting creation
2. Copying an automatically generated meeting does not work. The old agenda item is not copied over
3. The created agenda items do not have their presenter field set. The spec says that the info should be in the presenter field, but is instead in the author field (which isn't displayed). I don't know if this is fine like this or if it has to be displayed and hence added to the presenter field too

There's also some references for meeting minutes, meeting agenda, and meeting content in the API and stimulus controllers. Are they all still needed? For example:

- `/modules/meeting/lib/api/v3/meeting_minutes/meeting_minutes_representer.rb`
- `/frontend/src/stimulus/controllers/dynamic/meeting-content.controller.ts`

oliverguenther commented:
> Hi, there's still a few things that need to be looked at, apart from the minor inline comments:
> 
> 1. The "+ Meeting" menu still has an option for classic meeting creation

Classic. 🤦 

> 2. Copying an automatically generated meeting does not work. The old agenda item is not copied over

I can't reproduce this, could you elaborate on the steps that you used?

> 3. The created agenda items do not have their presenter field set. The spec says that the info should be in the presenter field, but is instead in the author field (which isn't displayed). I don't know if this is fine like this or if it has to be displayed and hence added to the presenter field too

Good catch! Fixed in the original migration

> 
> There's also some references for meeting minutes, meeting agenda, and meeting content in the API and stimulus controllers. Are they all still needed? For example:
> 
> * `/modules/meeting/lib/api/v3/meeting_minutes/meeting_minutes_representer.rb`
> * `/frontend/src/stimulus/controllers/dynamic/meeting-content.controller.ts`

Yes, can be removed. Did that and found some more. I hope to have caught all now

oliverguenther commented:
I added https://community.openproject.org/work_packages/63114 after discussing with Jens about the comments. We will:

- inform about the missing bits publicly (already part of the blog article, to be added to release notes)
- do not delete the journals for 16.0
- If demand is high, provide a migration, or more likely, provide an export of this history so we can attach it to the meeting
- delete the content journals in 17.0

mrmir commented on modules/meeting/app/models/meeting.rb:
```diff
@@ -41,17 +41,15 @@ class Meeting < ApplicationRecord
 
-  has_one :agenda, dependent: :destroy, class_name: "MeetingAgenda"
-  has_one :minutes, dependent: :destroy, class_name: "MeetingMinutes"
-  has_many :contents, -> { readonly }, class_name: "MeetingContent"
```
Since we're not deleting the contents, does all of this and the models etc need to stay too?

mrmir reviewed (approved):
👍🏽

oliverguenther committed:
Keep database records so we can provide a structured export

To be removed in
https://community.openproject.org/work_packages/63114/activity

mrmir commented on modules/meeting/app/models/meeting.rb:
```diff
@@ -50,8 +52,10 @@ class Meeting < ApplicationRecord
-  has_many :sections, dependent: :destroy, class_name: "MeetingSection"
-  has_many :agenda_items, dependent: :destroy, class_name: "MeetingAgendaItem"
+  has_many :agenda_items, dependent: :destroy, class_name: "MeetingAgendaItem", inverse_of: :meeting
+  has_many :sections, dependent: :delete_all, class_name: "MeetingSection"
```
Just noticed this while rebasing my PR, but why the change from `:destroy` to `:delete_all` for sections?
